master/internal/stream/trials.go

```diff
@@ -54,39 +53,48 @@ func (tm *TrialMsg) SeqNum() int64 {
 	return tm.Seq
 }
 
-func (tm *TrialMsg) UpsertMsg() *websocket.PreparedMessage {
+func (tm *TrialMsg) UpsertMsg(upsertFunc stream.UpsertFunc) interface{} {
 	wrapper := struct {
 		Trial *TrialMsg `json:"trial"`
 	}{tm}
+
+	if upsertFunc != nil {
+		return upsertFunc(tm)
+	}
 	return prepareMessageWithCache(wrapper, &tm.upsertCache)
 }
 
-func (tm *TrialMsg) DeleteMsg() *websocket.PreparedMessage {
+func (tm *TrialMsg) DeleteMsg(deleteFunc stream.DeleteFunc) interface{} {
 	deleted := strconv.FormatInt(int64(tm.ID), 10)
+
+	if deleteFunc != nil {
+		return deleteFunc(TrialsDeleteKey, deleted)
+	}
 	return newDeletedMsgWithCache(TrialsDeleteKey, deleted, &tm.deleteCache)
 }
 
 // TrialSubscriptionSpec is what a user submits to define a trial subscription.
 // determined:streamable
 type TrialSubscriptionSpec struct {
-	TrialIds      []int  `json:"trial_ids"`
-	ExperimentIds []int  `json:"experiment_ids"`
-	Since         int64  `json:"since"`
+	TrialIds      []int `json:"trial_ids"`
+	ExperimentIds []int `json:"experiment_ids"`
+	Since         int64 `json:"since"`
 }
 
-func TrialCollectStartupMsgs(known string, spec TrialSubscriptionSpec, ctx context.Context) (
-	[]*websocket.PreparedMessage, error,
+// TODO: refactor pls
+func TrialCollectStartupMsgs(known string, spec TrialSubscriptionSpec, ctx context.Context, upsertFunc stream.UpsertFunc, deleteFunc stream.DeleteFunc) (
+	[]interface{}, error,
 ) {
-	var out []*websocket.PreparedMessage
+	var out []interface{}
 
 	if len(spec.TrialIds) == 0 && len(spec.ExperimentIds) == 0 {
 		// empty subscription: everything known should be returned as deleted
-		out = append(out, newDeletedMsg(TrialsDeleteKey, known))
+		out = append(out, newDeletedInterface(TrialsDeleteKey, known, deleteFunc))
```

Comment:
nit: newDeletedInterface is unclear, maybe we could do something along the lines of:
- `newDeletedMsg()` -> `newDeletedPreparedMsg()` 
- `newDeletedInterface()` -> `newDeletedMsg()`